### PR TITLE
[runtime] Make mono_value_box use MonoError

### PR DESCRIPTION
**Pull request by lambdageek**
Mark it external only.  Runtime should use mono_value_box_checked.

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -3185,9 +3190,10 @@ ves_icall_InternalExecute (MonoReflectionMethod *method, MonoObject *this_arg, M
 				MonoClassField* field = mono_class_get_field_from_name (k, str);
 				if (field) {
 					MonoClass *field_klass =  mono_class_from_mono_type (field->type);
-					if (field_klass->valuetype)
-						result = mono_value_box (domain, field_klass, (char *)this_arg + field->offset);
-					else 
+					if (field_klass->valuetype) {
+						result = mono_value_box_checked (domain, field_klass, (char *)this_arg + field->offset, &error);
+						mono_error_set_pending_exception (&error);
```
Comment (kumpera) [inline mono/metadata/icall.c]: Shouldn't we return once we set a pending exception?

Comment (kumpera) [inline mono/metadata/icall.c]: NM, should work, but looks fishy.

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -519,7 +519,11 @@ mono_ftnptr_to_delegate (MonoClass *klass, gpointer ftn)
 
 		if (use_aot_wrappers) {
 			wrapper = mono_marshal_get_native_func_wrapper_aot (klass);
-			this_obj = mono_value_box (mono_domain_get (), mono_defaults.int_class, &ftn);
+			this_obj = mono_value_box_checked (mono_domain_get (), mono_defaults.int_class, &ftn, &error);
+			if (!is_ok (&error)) {
```
Comment (kumpera) [inline mono/metadata/marshal.c]: What if we made mono_error_set_pending_exception return a gboolean?

That way the code can become:
```
if (mono_error_set_pending_exception (&error))
  return NULL
````

Or maybe we could add a return_val_if_nok variant that sets a pending exception.

File: mono/metadata/object.c [hunk 3]
```diff
@@ -5796,23 +5803,42 @@ mono_string_new_wrapper (const char *text)
 MonoObject *
 mono_value_box (MonoDomain *domain, MonoClass *klass, gpointer value)
 {
-	MONO_REQ_GC_UNSAFE_MODE;
-
 	MonoError error;
+	MonoObject *result = mono_value_box_checked (domain, klass, value, &error);
+	mono_error_cleanup (&error); /* FIXME new API that doesn't swallow the error */
```
Comment (kumpera) [inline mono/metadata/object.c]: This function is now MONO_RT_EXTERNAL_ONLY, this FIXME can go away.

File: mono/metadata/icall.c (hunk 1 above)
Comment (lambdageek) [inline mono/metadata/icall.c]: Took me a while to figure out wth it's doing and it's only been a week: I'll add a comment that it's falling through to get to the `g_free`.

File: mono/mini/debugger-agent.c [hunk 4]
```diff
@@ -6062,7 +6063,8 @@ decode_value_internal (MonoType *t, int type, MonoDomain *domain, guint8 *addr,
 					g_free (vtype_buf);
 					return err;
 				}
-				*(MonoObject**)addr = mono_value_box (d, klass, vtype_buf);
+				*(MonoObject**)addr = mono_value_box_checked (d, klass, vtype_buf, &error);
```
Comment (kumpera) [inline mono/mini/debugger-agent.c]: Shit, this should be using mono_gc_wbarrier_generic_store.

**Comment by kumpera** [comment]
LGTM bunch of silly issues.
@monojenkins merge

**Pull request merged**